File: elfin_teleop/src/elfin_action_keyboard.cpp
#include "elfin_teleop/elfin_action_keyboard.h"

ElfinActionKeyboardNode::ElfinActionKeyboardNode(){
	// Initialize Variables
	ready = false;
	exec_time = 1;

	// ROS Infrastructure
	keys_sub = nh.subscribe(KEYS_TOPIC, 1, &ElfinActionKeyboardNode::keysCallback, this);
	joint_state_sub = nh.subscribe(JOINT_STATE_TOPIC, 1, &ElfinActionKeyboardNode::jointStateCallback, this);

	// Wait to get ready
	ROS_INFO("Waiting for joint_states...");
	while(!ready)
		ros::spinOnce();
	ROS_INFO("Received first joint_states message");

	// Initialize position
	joint_position = current_joint_state.position;

	// Action Library for ros_control
	robot_client = new Client("elfin_arm_controller/follow_joint_trajectory", true);

	ROS_INFO("Waiting for server...");
	robot_client->waitForServer();
	ROS_INFO("Connected to server");

	goal.trajectory.joint_names = joint_names;
	goal.trajectory.points.resize(1);
	goal.trajectory.points[0].positions.resize(6);
	goal.trajectory.points[0].velocities.resize(6);
	goal.trajectory.points[0].velocities = joint_velocity;
}

ElfinActionKeyboardNode::~ElfinActionKeyboardNode(){
	delete robot_client;
	nh.shutdown();
}

void ElfinActionKeyboardNode::keysCallback(const std_msgs::String::ConstPtr& msg){
	if (!ready)
		return;

	double i = 0.1;
	std::string key = msg->data.c_str();
	char key_char = (char) key[0];
	switch (key_char) {
	case 'q':
		joint_position[0] += i;
		break;
	case 'a':
		joint_position[0] -= i;
		break;
	case 'w':
		joint_position[1] += i;
		break;
	case 's':
		joint_position[1] -= i;
		break;
	case 'e':
		joint_position[2] += i;
		break;
	case 'd':
		joint_position[2] -= i;
		break;
	case 'r':
		joint_position[3] += i;
		break;
	case 'f':
		joint_position[3] -= i;
		break;
	case 't':
		joint_position[4] += i;
		break;
	case 'g':
		joint_position[4] -= i;
		break;
	case 'y':
		joint_position[5] += i;
		break;
	case 'h':
		joint_position[5] -= i;
		break;
	case 'z':
		joint_position = {0, 0, 0, 0, 0, 0};
		break;
	case 'm':
		joint_position = {1.57, 0.78, -1.57, 0, 0.78, 0.0};
	}

	limitJoints();

	//robot_client->cancelGoal();
	for (size_t i=0; i<ELFIN_DOF_JOINTS; i++) 
		goal.trajectory.points[0].positions[i] = joint_position[i];
	goal.trajectory.points[0].time_from_start = ros::Duration(exec_time);
	goal.trajectory.header.stamp = ros::Time::now();

	robot_client->sendGoal(goal);
	robot_client->waitForResult();

    // self.robot_client.cancel_goal()
    // self.g.trajectory.points = [JointTrajectoryPoint(positions=self.joint_position, velocities=self.joint_velocity, time_from_start=rospy.Duration(0.1))]
    // self.robot_client.send_goal(self.g)
    // self.robot_client.wait_for_result()	
}

void ElfinActionKeyboardNode::jointStateCallback(const sensor_msgs::JointState& msg){
	current_joint_state = msg;
	ready = true;
}

void ElfinActionKeyboardNode::limitJoints(){
	for (size_t i=0; i<ELFIN_DOF_JOINTS; i++){
		if (joint_position[i] > joint_upper_limit[i])
			joint_position[i] = joint_upper_limit[i];
		else if (joint_position[i] < joint_lower_limit[i])
			joint_position[i] = joint_lower_limit[i];
	}
}

int main(int argc, char** argv)
{
	ros::init(argc, argv, "elfin_action_keyboard");

	ElfinActionKeyboardNode node;
	ros::spin();

	return 0;
}

File: elfin_teleop/scripts/elfin_action_keyboard.py
#!/usr/bin/env python
import time
import copy
import rospy
import actionlib
from std_msgs.msg import String
from sensor_msgs.msg import JointState
from control_msgs.msg import *
from trajectory_msgs.msg import *

# ---------------- CLASS ----------------
class ElfinActionKeyboardNode():
    def __init__(self):
        print('ElfinActionKeyboardNode: initializing node')

        self.ready = False
        self.joint_states = None

        # Initialize Variables
        self.joint_upper_limit = [ 3.14,  2.35,  2.61,  3.14,  2.56,  3.14]
        self.joint_lower_limit = [-3.14, -2.35, -2.61, -3.14, -2.56, -3.14]
        self.joint_position = [0, 0, 0, 0, 0, 0]
        self.joint_velocity = [0, 0, 0, 0, 0, 0]
        self.joint_names = [
            'elfin_joint1', 
            'elfin_joint2', 
            'elfin_joint3', 
            'elfin_joint4', 
            'elfin_joint5', 
            'elfin_joint6'
        ]
        self.exec_time = 1

        # ROS Infrastructure
        topic = 'keys'
        self.sub_joint_cmd = rospy.Subscriber(topic, String, self.keys_callback)
        topic = 'joint_states'
        self.sub_joint_states = rospy.Subscriber(topic, JointState, self.joint_states_callback)
        
        # Wait to get ready
        print('Waiting for joint_states...')
        while (not self.ready):
            pass
        print('Received first joint_states message')

        self.joint_position = list(self.joint_states.position)
        
        # Action Library for ros_control
        namespace = 'elfin_arm_controller'
        self.robot_client = actionlib.SimpleActionClient(namespace + '/follow_joint_trajectory', FollowJointTrajectoryAction)

        print('Waiting for server...')
        self.robot_client.wait_for_server()
        print('Connected to server')

        self.g = FollowJointTrajectoryGoal()
        self.g.trajectory = JointTrajectory()
        self.g.trajectory.joint_names = self.joint_names

        # Initial Position (Optional)
        #self.joint_position = self.limit_joints(self.joint_position)
        #self.g.trajectory.points = [JointTrajectoryPoint(positions=self.joint_position, velocities=self.joint_velocity, time_from_start=rospy.Duration(2.0))]
        #self.robot_client.send_goal(self.g)
        #self.robot_client.wait_for_result()
        rospy.sleep(1)


    def keys_callback(self, msg):
        if not self.ready:
            return

        i = 0.1
        key = msg.data
        unlimited_joints = copy.copy(self.joint_position)
        if(key=="q"):
            unlimited_joints[0] += i
        elif(key=="a"):
            unlimited_joints[0] -= i
        
        elif(key=="w"):
            unlimited_joints[1] += i
        elif(key=="s"):
            unlimited_joints[1] -= i
     
        elif(key=="e"):
            unlimited_joints[2] += i
        elif(key=="d"):
            unlimited_joints[2] -= i
     
        elif(key=="r"):
            unlimited_joints[3] += i
        elif(key=="f"):
            unlimited_joints[3] -= i
     
        elif(key=="t"):
            unlimited_joints[4] += i
        elif(key=="g"):
            unlimited_joints[4] -= i
     
        elif(key=="y"):
            unlimited_joints[5] += i
        elif(key=="h"):
            unlimited_joints[5] -= i

        elif(key=="z"):
            unlimited_joints = [0, 0, 0, 0, 0, 0]
        elif(key=="m"):
            unlimited_joints = [1.57, 0.78, -1.57, 0, 0.78, 0.0]
        elif(key=="x"):
            unlimited_joints = list(self.joint_states.position)

        # Limit position
        self.joint_position = self.limit_joints(unlimited_joints)

        # Send goal
        self.robot_client.cancel_goal()
        self.g.trajectory.points = [JointTrajectoryPoint(positions=self.joint_position, velocities=self.joint_velocity, time_from_start=rospy.Duration(self.exec_time))]
        self.robot_client.send_goal(self.g)
        self.robot_client.wait_for_result()


    def joint_states_callback(self, msg):
        self.joint_states = msg
        self.ready = True


    def limit_joints(self, q):
        limited = copy.copy(q)
        for i in range(6):
            if q[i] > self.joint_upper_limit[i]:
                limited[i] = self.joint_upper_limit[i]
            elif q[i] < self.joint_lower_limit[i]:
                limited[i] = self.joint_lower_limit[i]
        return limited


if __name__ == '__main__':
    print('starting elfin_action_keyboard')
    rospy.init_node('elfin_action_keyboard', disable_signals=True)
    node = ElfinActionKeyboardNode()

    try:
        rospy.spin()
    except:
        print('caught exception')
    print('exiting')